docs/extra/optimize_function.rst

```diff
@@ -0,0 +1,297 @@
+  ...     else:
+  ...         print(f"Failed {method.__name__}: {res[3]}")
+  ...
+  Failed lmdif: improper input parameters
+  >>> print(tbl)
+      method           stat0                 x                  y
+    ---------- --------------------- ------------------ ------------------
+         minim  1.73832000295664e-11 1.9999999630648686  4.000644742396361
+    neldermead 4.932797434506576e-11 2.0000007686008545 3.9994204823962214
+    montecarlo 6.629393582604866e-12 1.9999997816942297  3.999564165982678
+
+In this particular case :py:func:`~sherpa.optmethods.optfcts.lmdif`
+failed, and this is because the ``to_optimize`` function returned None
+as its second argument. For the other cases we can see that they all
+found the minimum location.
+
+In order to use :py:func:`~sherpa.optmethods.optfcts.lmdif` we need
+the per-parameter statistic, which we can get with a small tweak::
+
+  def to_optimize2(args):
+      x = args[0]
+      y = args[1]
+      pa = (a - x) * (a - b)
+      pb = b * (y - x * x) * (y - x * x)
+      stat = pa * pa + pb * pb
+      return stat, [pa, pb]
+
+This lets us use `lmdif`::
+
+  >>> res2 = optfcts.lmdif(to_optimize2, start, lo, hi)
+  >>> res2[0]
+  True
+  >>> res2[1]
+  [2.        4.0000846]
+
+The `lmdif` optimizer is one of those that returns a number of
+elements in the "extra" dictionary::
+
+  >>> res2[4]
+  {'info': 0, 'nfev': 108, 'covar': array([[ 1.56250000e-02, -1.04361202e-01],
+         [-1.04361202e-01,  4.15098457e+03]]), 'num_parallel_map': 0}
+
+Optimizing a model
+------------------
+
+We can re-create the :ref:`Fitting the NormGauss1D and Lorentz1D models <fit_peaked_data_normgauss1d_lorentz1d>`
+section - at least for the :py:class:`~sherpa.models.basic.NormGauss1D` fit.
+
+The normalized gaussian model can be expressed as
+
+::
+
+  def ngauss(x, ampl, pos, fwhm):
+      term = 4 * np.log(2)
+      numerator = ampl * np.exp(-term * (x - pos) * (x - pos) / (fwhm * fwhm))
+      denominator = np.sqrt(np.pi / term) * fwhm
+      return numerator / denominator
+
+and the data we are going to fit is read from a file:
+
+  >>> d = np.loadtxt('test_peak.dat')
+  >>> x = d[:, 0]
+  >>> y = d[:, 1]
+
+In this case the "callback" routine we want to give the minimizer a
+routine which will apply a least-squares minimizer to compare the
+model to the data (the `x` and `y` values for the independent and
+dependent axes respectively).
+
+::
+
+  def cb(pars):
+      model = ngauss(x, pars[0], pars[1], pars[2])
+      delta = model - y
+      statval = (delta * delta).sum()
+      # Keep a record of the parameters we've visited
+      store.add_row([statval] + list(pars))
+      return statval, delta
+
+.. note::
+   This function would normally be written as a closure to ensure
+   that the use of global terms like `x`, `y`, `ngauss`, and `store` do not
+   cause problems.
+
+For the starting point and bounds I'm using to use a "guess" based on
```

Comment:
```suggestion
For the starting point and bounds I'm using a "guess" based on
```

Reply:
The rest of the test is written with "we do this". Now, you switch to "I". Intentional?

Reply:
I'm ambidextrous... We shoulf=d really scrub the documnetation for this sort of change in style - I do it a lot - but then I decide that consistency is a hobgoblin of minds that aren't knee deep in documentation and forget...